descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,307 @@
+        
+        #catalog quantities
+        self.z = z
+        possible_mag_fields = ('mag_{}_lsst',
+                               'mag_{}_sdss',
+                               'mag_{}_des',
+                              )
+        self.possible_mag_fields = [f.format(band) for f in possible_mag_fields]
+        self.band = band
+        
+        #z-bounds and binning
+        self.zlo = zlo
+        self.zhi = zhi
+        self.N_zbins = N_zbins
+        self.zbins = np.linspace(zlo,zhi,N_zbins+1)
+        self.filters = [(lambda z: (z > zlo) & (z < zhi), self.z)]
+
+        #validation data
+        self.validation_data = {}
+        possible_observations = data_info.keys()
+        self.observation = observation
+        #check for valid combinations
+        if len(observation)==0:
+            print ('Warning: no data file supplied, no observation requested; only catalog data will be shown')
+        else:
+            #check that observation is known
+            if not(observation in possible_observations):
+                raise ValueError('Observation {} not available'.format(observation))
+            else:
+                #fetch validation data
+                self.validation_data = self.get_validation_data(band, observation)
+                
+        #plotting variables
+        self.normed = normed
+        self.ncolumns = int(ncolumns)
+
+        #setup subplot configuration and get magnitude cuts for each plot
+        self.mag_lo,self.mag_hi = self.init_plots(m_lo,m_hi)
+        
+        #setup summary plot
+        self.summary_fig, self.summary_ax = plt.subplots(self.nrows,self.ncolumns,figsize=(figx_p,figy_p),sharex='col')
+        self.first_pass = True    #only plot validation data once
+
+        #other
+        self._other_kwargs = kwargs
+
+        return
+
+    def init_plots(self, mlo, mhi):
+        #get magnitude cuts based on validation data or default limits
+        mag_lo = self.validation_data.get('mag_lo', [float(m) for m in range(mhi, mlo+1)])
+        mag_hi = self.validation_data.get('mag_hi', [])
+
+        #setup plots and determine number of rows required for subplots
+        self.nplots = len(mag_lo)
+        self.nrows = (self.nplots+self.ncolumns-1)//self.ncolumns
+
+        #colors
+        self.colors = iter(default_colors)
+
+        return mag_lo, mag_hi
+
+    def get_catalog_data(self, gc, quantities, filters=[]):
+        
+        data = {}
+        if not gc.has_quantities(quantities):
+            return TestResult(skipped=True,summary='Missing requested quantities')
+        
+        data = gc.get_quantities(quantities,filters=filters)
+        #make sure data entries are all finite 
+        data = GCRQuery(*((np.isfinite, col) for col in data)).filter(data)
+
+        return data
+
+    def get_validation_data(self, band, observation):
```

Comment:
Refactor this function
```python
def get_validation_data(self, band, observation):
    data_args = possible_observations[observation]
    data_path = os.path.join(self.data_dir, data_args['filename_template'].format(band))

    if not os.path.exists(data_path):
        raise ValueError("{}-band data file {} not found".format(band, data_path))

    if not os.path.getsize(data_path):
        raise ValueError("{}-band data file {} is empty".format(band, data_path))
        
    data = np.loadtxt(data_path, unpack=True, usecols=data_args['usecols'], skiprows=data_args['skiprows'])

    validation_data = dict(zip(data_args['colnames'], data))
    validation_data['label'] = data_args['label']
    
    if 'mag_hi' not in validation_data:   
        validation_data['mag_hi'] = []
        validation_data['mag_lo'] = [float(m) for m in range(int(mag_hi_lim), int(mag_lo_lim)+1)]

    return validation_data
```

Reply:
OK this is much slicker than my code. I didn't realize you could do the dict(zip(... trick. That's pretty neat.  Anyway, done.